cloudevents/pkg/reconciler/reconciler.go

```diff
@@ -0,0 +1,145 @@
+)
+
+type CDEventType string
+
+const (
+//// Environment Annotation
+//cdEnv = "tekton.dev/cd.environment.v1"
+//// Environment Annotation Values
+//cdEnvCreated = "created"
+//cdEnvModified = "modified"
+//cdEnvDeleted = "deleted"
+//
+//// Service Annotation
+//cdSvc = "tekton.dev/cd.service.v1"
+//// states
+//cdSvcStarted = "deployed"
+//cdSvcUpgraded = "upgraded"
+//cdSvcRolledback = "rolledback"
+//cdSvcRemoved = "removed"
+//
+//// TaskRun Annotation
+//cdTR = "tekton.dev/cd.taskrun.v1"
+//// states
+//cdTRStarted = "started"
+//cdTRFinished = "finished"
+//cdTRQueued = "queued"
+//
+//// Repository Annotation
+//cdRepo = "tekton.dev/cd.repository.v1"
+//// states
+//cdRepoCreated = "created"
+//cdRepoModified = "modified"
+//cdRepoDeleted = "deleted"
+
+//// PipelineRun Annotations
+//cdPipelineRunID = "cd.tekton.dev/pipelinerun.v1/id"
+//// states
+//cdPRStarted = "started"
+//cdPRFinished = "finished"
+//cdPRQueued = "queued"
+//
+//// PipelineRun events
+//PipelineRunStartedEventV1 CDEventType = "cd.pipelinerun.started.v1"
+//PipelineRunFinishedEventV1 CDEventType = "cd.pipelinerun.finished.v1"
+//PipelineRunQueuedEventV1 CDEventType = "cd.pipelinerun.queued.v1"
+)
+
+func (t CDEventType) String() string {
+	return string(t)
+}
+
+type Reconciler struct {
+	cloudEventClient cloudevent.CEClient
+	tracker          tracker.Interface
+
+	taskRunLister listers.TaskRunLister
+	snooze        func(kmeta.Accessor, time.Duration)
+}
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (c *Reconciler) ReconcileKind(ctx context.Context, pr *v1beta1.PipelineRun) kreconciler.Event {
+	logger := logging.FromContext(ctx)
+	ctx = cloudevent.ToContext(ctx, c.cloudEventClient)
+
+	// Read the initial condition
+	before := pr.Status.GetCondition(apis.ConditionSucceeded)
+
+	if !pr.HasStarted() && !pr.IsDone() { //&& !pr.IsPending() {
+		afterCondition := pr.Status.GetCondition(apis.ConditionSucceeded)
+		events.Emit(ctx, nil, afterCondition, pr)
+
+		// We already sent an event for start, so update `before` with the current status
+		before = pr.Status.GetCondition(apis.ConditionSucceeded)
+	}
+
+	if pr.IsDone() {
+		c.finishReconcileUpdateEmitEvents(ctx, pr, before, nil)
+		return nil
+	}
+
+	if pr.IsCancelled() {
+		c.finishReconcileUpdateEmitEvents(ctx, pr, before, nil)
+		return nil
+	}
+
+	if err := c.tracker.TrackReference(tracker.Reference{
+		APIVersion: v1beta1.SchemeGroupVersion.String(),
+		Kind:       "TaskRun",
+		Namespace:  pr.GetNamespace(),
+		Name:       pr.GetName(),
+	}, pr); err != nil {
+		logger.Errorf("Failed to create tracker for TaskRuns for PipelineRun %s: %v", pr.Name, err)
+		c.finishReconcileUpdateEmitEvents(ctx, pr, before, err)
+		return nil
+	}
+
+	defer func() {
+		if pr.Status.StartTime == nil {
+			return
+		}
+		// Compute the time since the task started.
+		elapsed := time.Since(pr.Status.StartTime.Time)
+		// Snooze this resource until the timeout has elapsed.
+		c.snooze(pr, pr.GetTimeout(ctx)-elapsed)
```

Comment:
The cloud event controller should not take care of implementing the timeout - that is done by the main controller.

Reply:
removed the defer statement which takes care of the timeout